core/commands/files/files.go

```diff
@@ -0,0 +1,697 @@
+		node, err := req.InvocContext().GetNode()
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		src, err := checkPath(req.Arguments()[0])
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+		dst, err := checkPath(req.Arguments()[1])
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		nd, err := getNodeFromPath(req.Context(), node, src)
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		err = mfs.PutNode(node.FilesRoot, dst, nd)
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+	},
+}
+
+func getNodeFromPath(ctx context.Context, node *core.IpfsNode, p string) (*dag.Node, error) {
+	switch {
+	case strings.HasPrefix(p, "/ipfs/"):
+		np, err := path.ParsePath(p)
+		if err != nil {
+			return nil, err
+		}
+
+		return core.Resolve(ctx, node, np)
+	default:
+		fsn, err := mfs.Lookup(node.FilesRoot, p)
+		if err != nil {
+			return nil, err
+		}
+
+		return fsn.GetNode()
+	}
+}
+
+type Object struct {
+	Hash           string
+	Size           uint64
+	CumulativeSize uint64
+	Blocks         int
+}
+
+type FilesLsOutput struct {
+	Entries []mfs.NodeListing
+}
+
+var FilesLsCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "List directories",
+		ShortDescription: `
+List directories.
+
+Examples:
+
+    $ ipfs files ls /welcome/docs/
+    about
+    contact
+    help
+    quick-start
+    readme
+    security-notes
+
+    $ ipfs files ls /myfiles/a/b/c/d
+    foo
+    bar
+`,
+	},
+	Arguments: []cmds.Argument{
+		cmds.StringArg("path", true, false, "path to show listing for"),
+	},
+	Options: []cmds.Option{
+		cmds.BoolOption("l", "use long listing format"),
+	},
+	Run: func(req cmds.Request, res cmds.Response) {
+		path, err := checkPath(req.Arguments()[0])
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		nd, err := req.InvocContext().GetNode()
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		fsn, err := mfs.Lookup(nd.FilesRoot, path)
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		switch fsn := fsn.(type) {
+		case *mfs.Directory:
+			listing, err := fsn.List()
+			if err != nil {
+				res.SetError(err, cmds.ErrNormal)
+				return
+			}
+			res.SetOutput(&FilesLsOutput{listing})
+			return
+		case *mfs.File:
+			parts := strings.Split(path, "/")
+			name := parts[len(parts)-1]
+			out := &FilesLsOutput{[]mfs.NodeListing{mfs.NodeListing{Name: name, Type: 1}}}
+			res.SetOutput(out)
+			return
+		default:
+			res.SetError(errors.New("unrecognized type"), cmds.ErrNormal)
+		}
+	},
+	Marshalers: cmds.MarshalerMap{
+		cmds.Text: func(res cmds.Response) (io.Reader, error) {
+			out := res.Output().(*FilesLsOutput)
+			buf := new(bytes.Buffer)
+			long, _, _ := res.Request().Option("l").Bool()
+
+			for _, o := range out.Entries {
+				if long {
+					fmt.Fprintf(buf, "%s\t%s\t%d\n", o.Name, o.Hash, o.Size)
```

Comment:
You might want to try [tabwriter](https://godoc.org/text/tabwriter) next time :)